### PR TITLE
update generator aspect

### PR DESCRIPTION
**Pull request by NitsanCohen770**
### Pull Request: update-generator-aspect

**Summary:**

- **Introduced Generics:** Improved type flexibility by introducing generics for both the `WorkspaceOptions` and `ComponentContext`. This change allows users to define specific types for the `aspect` and `template` fields without altering the core structure.
  
- **Conditionally Populate `.gitignore`:** Updated the starter template logic to conditionally populate the `.gitignore` file based on the `skipGit` flag in the `WorkspaceContext`.

- **Enhanced Component Templates:** Updated the component templates for the starter and the component-generator to reflect the latest patterns, ensuring consistency and adaptability.
  
- **Legacy Templates Update:** Legacy templates have been set to hidden, and all legacy workspace templates have been removed from older core environments.
  
**A Bit of Humor:**
It's been a unique challenge creating templates for templates! It's almost like trying to explain a joke without ruining the punchline...almost.

**Important Notes:**

- **Documentation Update:** Please hold off on merging this PR until the documentation is ready. Expect it very soon!
  
**Motivation:**

The primary catalyst behind introducing generics was observing the significant type duplication across the codebase. These duplications were not only cumbersome to manage and update but also led to confusion. By consolidating similar types through generics, we've reduced redundancy and streamlined the type architecture.

Moreover, it was concerning that when developers use Bit, they predominantly encounter legacy templates. This doesn't showcase the capabilities of our new environments and might lead to a perception lag. Hence, with this update, we aim to pivot developers towards the more recent and efficient templates, providing them a taste of Bit's cutting-edge capabilities right out of the gate.

### CODE REVIEW

File: scopes/generator/generator/templates/component-generator/files/component-template.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+      readonly name = '${name}',
+      readonly description = 'a template for ${name} components',
+      readonly hidden = false
+    ) {}
+  
+    generateFiles(context: ComponentContext): ComponentFile[] {
+      return [
+        indexFile(context),
+        compositionFile(context),
+        componentFile(context),
```
Comment (GiladShoham) [inline scopes/generator/generator/templates/component-generator/files/component-template.ts:42]: What about docs file here?

File: scopes/generator/generator/templates/component-generator/index.ts [hunk 2]
```diff
@@ -1,31 +1,41 @@
 import { ComponentContext, ComponentTemplate } from '@teambit/generator';
 import { indexFile } from './files/index';
-import { aspectFile } from './files/aspect-file';
-import { docsFile } from './files/docs-file';
```
Comment (GiladShoham) [inline scopes/generator/generator/templates/component-generator/index.ts:4]: What about docs here?

File: scopes/generator/generator/templates/component-generator/index.ts [hunk 3]
```diff
@@ -1,31 +1,41 @@
...
+      },
       {
-        relativePath: `${context.name}.aspect.ts`,
-        content: aspectFile(context),
+        relativePath: `./files/composition-file.ts`,
+        content: generateCompositionFileContent(),
       },
       {
-        relativePath: `${context.name}.docs.mdx`,
-        content: docsFile(),
```
Comment (GiladShoham) [inline scopes/generator/generator/templates/component-generator/index.ts:24]: What about the docs?

File: e2e/harmony/custom-env.e2e.ts [hunk 4]
```diff
@@ -293,7 +293,7 @@ describe('custom env', function () {
       before(() => {
         helper.scopeHelper.reInitLocalScope();
         helper.scopeHelper.addRemoteScope();
-        helper.command.create('aspect', 'comp1', `--env ${envId}@0.0.1`);
+        helper.command.create('starter', 'comp1', `--env ${envId}@0.0.1`);
```
Comment (davidfirst) [inline e2e/harmony/custom-env.e2e.ts:296]: Please return the aspect template, I love it :)

File: e2e/harmony/new.e2e.ts [hunk 5]
```diff
@@ -27,29 +27,32 @@ describe('new command', function () {
...
     });
     after(() => {
       helper.scopeHelper.removeRemoteScope(undefined, true);
     });
-    it('should generate a new workspace, import components and add them as new with no issues', () => {
+    it('should generate a new workspace, import components and add with no issues', () => {
       const wsPath = path.join(helper.scopes.localPath, 'my-workspace');
-      const status = helper.command.statusJson(wsPath);
-      expect(status.newComponents).to.have.lengthOf(1);
+      const list = JSON.parse(helper.command.runCmd(`bit list --json`, wsPath));
```
Comment (davidfirst) [inline e2e/harmony/new.e2e.ts:54]: can you change it back to check the status of the newComponents?

File: scopes/community/community/community.main.runtime.ts [hunk 6]
```diff
@@ -20,14 +27,20 @@ export class CommunityMain {
...
   static defaultConfig: CommunityWorkspaceConfig = {
     communityDomain: BASE_COMMUNITY_DOMAIN,
     docsDomain: BASE_DOCS_DOMAIN,
   };
-  static async provider(_deps, config: CommunityWorkspaceConfig) {
-    return new CommunityMain(config);
+
+  static async provider(_deps: [GeneratorMain], config: CommunityWorkspaceConfig) {
+    const [generator] = _deps;
+    const community = new CommunityMain(config, generator);
```
Comment (davidfirst) [inline scopes/community/community/community.main.runtime.ts:39]: no need for this here

File: scopes/generator/generator/component-template.ts [hunk 7]
```diff
@@ -1,23 +1,12 @@
...
-}
-
-export interface ComponentContext {
+/**
+ * ComponentContextOptions describes the shared properties between command and context
+ * related to component initialization and configuration.
+ *
+ * @template TComponentID - Represents the data type for the component ID properties.
+ */
+export interface ComponentTemplateOptions<TComponentID = ComponentID> {
```
Comment (davidfirst) [inline scopes/generator/generator/component-template.ts:9]: Please change this interface to have only the shared props of the two that implement it. 
No need for the generic here. Each can have its own (string/ComponentId).

**Review by davidfirst — CHANGES_REQUESTED** [review]

File: e2e/harmony/new.e2e.ts (hunk 5 above)
Comment (NitsanCohen770) [reply e2e/harmony/new.e2e.ts:54]: helper.command.expectStatusToNotHaveIssues(wsPath) run bit status anyway and I am checking the same check on the bit list as you can see

File: scopes/generator/generator/templates/component-generator/files/component-template.ts (hunk 1 above)
Comment (NitsanCohen770) [reply scopes/generator/generator/templates/component-generator/files/component-template.ts:42]: @GiladShoham I followed the pattern we have in our component generator for React

File: scopes/generator/generator/templates/component-generator/index.ts (hunk 2 above)
Comment (NitsanCohen770) [reply scopes/generator/generator/templates/component-generator/index.ts:4]: @GiladShoham I followed the pattern we have in our component generator for React

File: scopes/generator/generator/templates/component-generator/index.ts (hunk 3 above)
Comment (NitsanCohen770) [reply scopes/generator/generator/templates/component-generator/index.ts:24]: @GiladShoham I followed the pattern we have in our component generator for React

**Review by davidfirst — DISMISSED** [review]

**Review by davidfirst — APPROVED** [review]

**Pull request merged**